File: src/nbla/function/generic/leaky_relu.cpp
// Copyright (c) 2017 Sony Corporation. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/** LeakyReLU
 */

#include <nbla/array.hpp>
#include <nbla/function/leaky_relu.hpp>
#include <nbla/variable.hpp>

#include <algorithm>

namespace nbla {

NBLA_REGISTER_FUNCTION_SOURCE(LeakyReLU, float, bool);

template <typename T>
void LeakyReLU<T>::setup_impl(const Variables &inputs,
                              const Variables &outputs) {
  outputs[0]->reshape(inputs[0]->shape(), true);
  if (inplace_) {
    NBLA_CHECK(
        alpha_ > 0, error_code::value,
        "Alpha must be greater than zero with inplace option being true.");
    outputs[0]->data()->set_array(inputs[0]->data()->array());
  }
}

template <class T>
void LeakyReLU<T>::forward_impl(const Variables &inputs,
                                const Variables &outputs) {
  const T *x = inputs[0]->get_data_pointer<T>(this->ctx_);
  T *y = outputs[0]->cast_data_and_get_pointer<T>(this->ctx_, !inplace_);
  for (int s = 0; s < inputs[0]->size(); s++) {
    T x_s = x[s];
    if (x_s > (T)0.)
      y[s] = x_s;
    else
      y[s] = alpha_ * x_s;
  }
}
template <typename T, bool accum>
void leaky_relu_backward_cpu(int size, float alpha, T *dx, const T *dy,
                             const T *x) {
  for (int s = 0; s < size; ++s) {
    if (accum) {
      if (x[s] > (T)0.)
        dx[s] += dy[s];
      else
        dx[s] += alpha * dy[s];
    } else {
      if (x[s] > (T)0.)
        dx[s] = dy[s];
      else
        dx[s] = alpha * dy[s];
    }
  }
}

template <class T>
void LeakyReLU<T>::backward_impl(const Variables &inputs,
                                 const Variables &outputs,
                                 const vector<bool> &propagate_down,
                                 const vector<bool> &accum) {
  if (!propagate_down[0]) {
    return;
  }
  const T *x = inputs[0]->get_data_pointer<T>(this->ctx_);
  T *dx = inputs[0]->cast_grad_and_get_pointer<T>(this->ctx_, !accum[0]);
  const T *dy = outputs[0]->get_grad_pointer<T>(this->ctx_);
  if (accum[0])
    leaky_relu_backward_cpu<T, true>(inputs[0]->size(), alpha_, dx, dy, x);
  else
    leaky_relu_backward_cpu<T, false>(inputs[0]->size(), alpha_, dx, dy, x);
}
}

File: python/src/nnabla/backward_function/leaky_relu.py
# Copyright 2019,2020,2021 Sony Corporation.
# Copyright 2021 Sony Group Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import nnabla.functions as F

from .utils import no_grad


def leaky_relu_backward(grad_inputs, inputs, input_shapes, outputs, output_shapes, alpha=0.1, inplace=False):
    """
    Args:
      grad_inputs (list of :obj:`nnabla.Variable`): Propagated grads to this backward function.
      inputs (list of :obj:`nnabla.Variable` and None): Input Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      input_shapes (list of tuple of :obj:`int`): Input shapes of the forward function.
          The shapes of the inputs in which None is set can be passed.
      outputs (list of :obj:`nnabla.Variable` and None): Output Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      output_shapes (list of tuple of :obj:`int`): Output shapes of the forward function.
          The shapes of the outputs in which None is set can be passed.
      kwargs (dict of arguments): Dictionary of the corresponding function arguments.

    Return:
      list of Variable: Return the gradients wrt inputs of the corresponding function.
    """
    dy = grad_inputs[0]
    x0 = inputs[0]

    # Keep the consistency with the grad dependency defined in C++ LeakyReLU.
    # Inplace option is ignored because it was obsoleted. See leaky_relu.hpp
    if alpha >= 0:
        # grad_depends_output_data=True
        x0 = outputs[0]
        # else grad_depends_input_data=True

    m0 = F.greater_scalar(x0, 0)  # result is same even if inplace or not
    m1 = 1 - m0
    m0 = no_grad(m0)
    m1 = no_grad(m1)
    dx0 = dy * (m0 + alpha * m1)
    return dx0
